### PR TITLE
First try at authentication

### PR DESCRIPTION
**Pull request by scheul93**

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -4,17 +4,35 @@ import Header from './Header/Header.js'
...
-        return (
-            <div className="App">
-                <Header />
-                <NavBar />
-                <div className="main"></div>
-                <Footer />
-            </div>
-        );
-    }
+	constructor(){
```
Comment (MeKamm74) [inline src/App.js:11]: nit: You have a no space between the parens and the bracket, but below there is one. I'm pretty sure it's from my code, hence the discrepancy, but we should probably pick one.

Comment (MeKamm74) [reply src/App.js:11]: Looking through your code, seems like you favor less unnecessary spaces in general? I'm just trying to get a feel for the style.

File: src/NavBar/NavBar.js [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
 import './NavBar.css';
 
-const NavBar = function () {
+const NavBar = function (props) {
 	return (
 		<div className="nav-bar">
 			<a className="nav-bar__item">Submit a book</a>
 			<a className="nav-bar__item">Submissions</a>
 			<a className="nav-bar__item">Rate a book</a>
+			<SignIn isAuthenticated={props.isAuthenticated} />
```
Comment (MeKamm74) [inline src/NavBar/NavBar.js:11]: So here we're passing props.isAuthenticated into SignIn. 

Do components in react automatically get access to properties that parents pass in? It doesn't look like you declare that property anywhere in the SignIn, you just access it through props.

File: src/SignIn/SignIn.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import React, { Component } from 'react';
+import authService from '../firebase/authService';
+
+class SignIn extends Component {
+    constructor() {
+        super();
```
Comment (MeKamm74) [inline src/SignIn/SignIn.js]: Does this call a default constructor from the base Component class? Should we always be calling this?

File: src/SignIn/SignIn.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    signOut() {
+        authService.signOut();
+    }
+
+    render() {
+        if(this.props.isAuthenticated){
+            return <button onClick={this.signOut.bind(this)}>Sign Out</button>
+        }
+
+        return <button onClick={this.signIn.bind(this)}>Sign In</button>
```
Comment (MeKamm74) [inline src/SignIn/SignIn.js]: to clarify, this is binding the signIn function to this button right? so signIn is called when the button is clicked?

File: src/firebase/authService.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        auth = firebase.auth();
+        auth.getRedirectResult().then(result => {
+            if(result && result.user) {
+                userService.addUser(result.user);
+            }
+        })
+
+        firebase.auth().onAuthStateChanged(user => {
+            if(user) {
+                console.log(user);
```
Comment (MeKamm74) [inline src/firebase/authService.js]: Why are we doing this? Just for troubleshooting?

File: src/firebase/authService.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import firebase from '@firebase/app';
+import '@firebase/auth';
+import userService from './userService';
+
+const authService = function() {
+    let auth = null;
+
+    function init() {
```
Comment (MeKamm74) [inline src/firebase/authService.js:8]: What's the difference between init() and constructor()?

File: src/firebase/authService.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    return {
+        init,
+        getCurrentUser,
+        signIn,
+        signOut
+    }
+}
+
+const instance = authService();
```
Comment (MeKamm74) [inline src/firebase/authService.js:40]: Any reason for this instead of 
`export default authservice;`

File: src/index.js [hunk 8]
```diff
@@ -2,7 +2,15 @@ import React from 'react';
 import ReactDOM from 'react-dom';
 import './index.css';
 import App from './App';
-import registerServiceWorker from './registerServiceWorker';
+// import registerServiceWorker from './registerServiceWorker';
```
Comment (MeKamm74) [inline src/index.js:5]: 1) I have no idea what registerServiceWorker does, I think it was in the boilerplate, do you know?
2) If it's really useless then you can just delete the comments.

**Comment by MeKamm74** [comment]
Most of these comments are just questions about react. And a little on style.

File: src/index.js (hunk 8 above)
Comment (scheul93) [reply src/index.js:5]: Pretry sure service workers are related to pwas and loading content while offline. I left it in so it we choose to make this a pwa, it's already built in to the boilerplate

File: src/firebase/authService.js (hunk 7 above)
Comment (scheul93) [reply src/firebase/authService.js:40]: This method makes the auth service a singleton. There is only one instance for the entire app.

In other components, you could have different instances of a component throughout the app.

File: src/firebase/authService.js (hunk 6 above)
Comment (scheul93) [reply src/firebase/authService.js:8]: I use init for actual logic to initialize the component.

I reserve the construct for initializing state.

Somewhere on the interwebs there is an explanation. I'm on mobile now, so I can link to it later

File: src/firebase/authService.js (hunk 5 above)
Comment (scheul93) [reply src/firebase/authService.js]: My b

File: src/SignIn/SignIn.js (hunk 4 above)
Comment (scheul93) [reply src/SignIn/SignIn.js]: Yep. So onClick is a built in React thing that you can pass a callback to.

In React you also have to bind functions, so I just did it here.

File: src/SignIn/SignIn.js (hunk 3 above)
Comment (scheul93) [reply src/SignIn/SignIn.js]: Honestly, not sure if this is necessary

File: src/NavBar/NavBar.js (hunk 2 above)
Comment (scheul93) [reply src/NavBar/NavBar.js:11]: Yeah, so parents pass props to children as attributes on elements (or at least, You can think of it like that syntactically). The Nav bar component reads in the props variable and then this component can pass those props to its children

**Review by MeKamm74 — APPROVED** [review]

**Pull request merged**